core/src/components/datetime/utils/manipulation.ts

```diff
@@ -472,25 +472,64 @@ export const getClosestValidDate = ({
   const copyParts = clampDate({ ...refParts, dayOfWeek: undefined }, minParts, maxParts);
 
   if (monthValues !== undefined) {
-    copyParts.month = findClosestValue(month, monthValues);
```

Comment:
We need to account for the already adjusted year values first like in my original example. We don't need to adjust the month by the min/max bounds if the year isn't the same as the min/max year. We'll want to do the same for the day logic too.

Test case:
```js
// April 2, 2020 9:16 AM
const refParts = { month: 4, day: 2, year: 2022, hour: 9, minute: 16 };
// September 10, 2021 10:10 AM
const minParts = { month: 9, day: 10, year: 2021, hour: 10, minute: 10 };
// September 10, 2023 11:15 AM
const maxParts = { month: 9, day: 10, year: 2023, hour: 11, minute: 15 };

// September 10, 2021 10:15 AM
const expected = {
  month: 4,
  day: 2,
  year: 2022,
  hour: 9,
  minute: 16,
  ampm: 'am',
  dayOfWeek: undefined,
};

expect(
  getClosestValidDate({
    refParts,
    monthValues: [4, 9, 11],
    dayValues: [2, 10, 12, 13, 14],
    yearValues: [2020, 2021, 2023],
    hourValues: [9, 10, 11],
    minuteValues: [10, 15, 16],
    minParts,
    maxParts,
  })
).toEqual(expected);
```